server-common/src/main/java/org/apache/kafka/server/purgatory/DelayedOperationPurgatory.java

```diff
@@ -86,6 +86,7 @@ public DelayedOperationPurgatory(String purgatoryName,
         this.purgeInterval = purgeInterval;
         this.reaperEnabled = reaperEnabled;
         this.timerEnabled = timerEnabled;
+        this.expirationReaper = new ExpiredOperationReaper();
```

Comment:
that seems error-prone to me, since the issue is caused by an incorrect initialization order. Perhaps we should use anonymous class to ensure the initialization order. for example:
```java
        this.expirationReaper = new ShutdownableThread("ExpirationReaper-" + brokerId + "-" + purgatoryName, false) {
            @Override
            public void doWork() {
                try {
                    advanceClock(200L);
                } catch (InterruptedException ie) {
                    throw new RuntimeException(ie);
                }
            }
        };
```

Reply:
Apologies for unclear comment. My point was that the `expirationReaper` initialization must happen after `brokerId` and `purgatoryName` are initialized. The dependency is implicit, which means we might inadvertently re-introduce a bug during future refactoring. Using an anonymous class could eliminate the happen-before constraint as it can directly reference the input arguments. 

Alternatively, we could add comments to explicitly remind developers that `expirationReaper` must be initialized after `brokerId` and `purgatoryName`

Reply:
Make sense. I have added the comment for now. Though the other suggestion was also valid.